### PR TITLE
Add surface albedo, accumulated snow melt, and total soil moisture output in RRFS

### LINKED ISSUE #904: Add accumulated snow melt and surface albedo output in RRFS
Accumulated snow melt (SNOM) and surface albedo are needed in RRFS.  Also, we need to encode the total soil moisture as CISOILM.

### LINKED ISSUE #906: Bug fix for blowing snow effect
During the beta evaluation, the Aviation Weather Center has noted that the visibility reductions associated with blowing snow in RRFS are excessive in forested areas.  The diagnostic will be modified based on this feedback.

### PR DESCRIPTION
**Pull request by EricJames-NOAA**
This PR enables calculation and output of surface albedo, accumulated snow melt, and total soil moisture from RRFS.

The code was tested for RRFS-A (3km North America domain) on Jet.  The only differences in output were for these variables.

### CODE REVIEW

File: sorc/ncep_post.fd/CALVIS_GSD.f [hunk 1]
```diff
@@ -360,7 +360,7 @@ SUBROUTINE CALVIS_GSD(CZEN,VIS)
         endif
 
 !  Calculation of visibility based on hydrometeor and aerosols.  (RH effect not yet included.)
-        VIS(I,J)=MIN(90.,CONST1/(BETAV+extcof55(i,j,lm)))      ! max of 90km
+        VIS(I,J)=MIN(90.,CONST1/BETAV)      ! max of 90km
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/CALVIS_GSD.f:365]: @EricJames-NOAA Can you add a change log at the beginning of this subroutine?

File: sorc/ncep_post.fd/SURFCE.f [hunk 2]
```diff
@@ -720,6 +721,32 @@ SUBROUTINE SURFCE
          endif
       ENDIF
 !
+!     TOTAL SOIL MOISTURE
+      IF (IGET(713)>0) THEN
+!$omp parallel do private(i,j)
+         DO J=JSTA,JEND
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/SURFCE.f:727]: @EricJames-NOAA Could you adapt the settings of UPP ID 036 (soil moisture) as:
```
 DO J=JSTA,JEND
           DO I=ISTA,IEND
             IF(SMSTOT(I,J)/=SPVAL) THEN
               IF(SM(I,J) > SMALL .AND. SICE(I,J) < SMALL) THEN
                 GRID1(I,J) = 1000.0  ! TEMPORY FIX TO MAKE SURE SMSTOT=1 FOR WATER
               ELSE
                 GRID1(I,J) = SMSTOT(I,J)
               END IF
             ELSE
               GRID1(I,J) = 1000.0
             ENDIF
           ENDDO
         ENDDO
```
So the new soil moisture ID can be used for all UFS based applications.

Comment (EricJames-NOAA) [reply sorc/ncep_post.fd/SURFCE.f:727]: Shouldn't we preserve ID 036 for output of SOILMOI in kg/m3 in case anyone needs that?

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/SURFCE.f:727]: @EricJames-NOAA Yes, We can keep ID 036 to output SOILMOI for non-UFS based models. After your PR is merged for RRFS, I will update total soil moisture grib2 name for rest of UFS based applications.

Comment (EricJames-NOAA) [reply sorc/ncep_post.fd/SURFCE.f:727]: @WenMeng-NOAA ok.  Sorry if I don't understand...what changes should I make in this PR?

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/SURFCE.f:727]: @EricJames-NOAA Could you add the same settings of ID 036 in this new ID block? such as:
```
DO J=JSTA,JEND
           DO I=ISTA,IEND
             IF(SMSTOT(I,J)/=SPVAL) THEN
               IF(SM(I,J) > SMALL .AND. SICE(I,J) < SMALL) THEN
                 GRID1(I,J) = 1000.0  ! TEMPORY FIX TO MAKE SURE SMSTOT=1 FOR WATER
               ELSE
                 GRID1(I,J) = SMSTOT(I,J)
               END IF
             ELSE
               GRID1(I,J) = 1000.0
             ENDIF
           ENDDO
         ENDDO
```

Comment (EricJames-NOAA) [reply sorc/ncep_post.fd/SURFCE.f:727]: @WenMeng-NOAA I'm not sure we need all those checks if the SMSTOT is really in kg/m2.  In that case, the current code for ID 713 retains SPVAL over all water points, and only has values over land.  I don't think it makes sense to assign column integrated soil moisture to a numeric value over water, I think we should just leave them as missing values.

If SMSTOT is in kg/m3, then other applications can use the existing ID 036.

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/SURFCE.f:727]: @EricJames-NOAA Okay, let's keep RRFS as your current setting. The "GRID1(I,J) = 1000.0" is a legacy setting for GFS products. I will evaluate GFS products, coordinate with users and submit separate PR for gfs modification only if needed later. Please let me know if this approach works for you.

Comment (EricJames-NOAA) [reply sorc/ncep_post.fd/SURFCE.f:727]: @WenMeng-NOAA OK, thank you, that sounds like a good plan to me.

**Comment by WenMeng-NOAA** [comment]
@EricJames-NOAA Can you provide me the RRFS model files for my UPP standalone test?

**Comment by EricJames-NOAA** [comment]
> @EricJames-NOAA Can you provide me the RRFS model files for my UPP standalone test?

@WenMeng-NOAA I have some files here on Hera:

/scratch1/BMC/wrfruc/ejames/wen_snom_cisoilm/

**Comment by WenMeng-NOAA** [comment]
@EricJames-NOAA Please sync your branch with the latest UPP develop branch.

File: sorc/ncep_post.fd/INITPOST_NETCDF.f [hunk 3]
```diff
@@ -3106,13 +3112,31 @@ SUBROUTINE INITPOST_NETCDF(ncid2d,ncid3d)
...
+            acsnom(i,j) = spval
+          endif
+        enddo
+      enddo
 !$omp parallel do private(i,j)
       do j=jsta_2l,jend_2u
         do i=ista_2l,iend_2u
 !          smstav(i,j) = spval    ! GFS does not have soil moisture availability
-!          smstot(i,j) = spval    ! GFS does not have total soil moisture
+          smstot(i,j) = spval    ! GFS does not have total soil moisture
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/INITPOST_NETCDF.f]: @EricJames-NOAA Since smstot is initialized in ALLOCATE_ALL.f, you might revert the change in this line.

File: parm/fv3lam_rrfs.xml [hunk 4]
```diff
@@ -399,8 +399,8 @@
        </param>
 
        <param>
-       <shortname>SOILM_ON_DEPTH_BEL_LAND_SFC</shortname>
-       <pname>SOILMOI</pname>
+       <shortname>CISOILM_ON_DEPTH_BEL_LAND_SFC</shortname>
```
Comment (WenMeng-NOAA) [inline parm/fv3lam_rrfs.xml:402]: @EricJames-NOAA The total soil moisture is generated as RRFS product as:
```
790:5274112429:d=2024022718:CISOILM:0-0 m below ground
``` 
There might be two soil level enters needed. 
The GFS is labeling as "0-2m" so enters in GFS xml are:
```
       <scale_fact_fixed_sfc1>2</scale_fact_fixed_sfc1>
       <level>0.</level>
       <scale_fact_fixed_sfc2>2</scale_fact_fixed_sfc2>
       <level2>200.</level2>
```

Comment (EricJames-NOAA) [reply parm/fv3lam_rrfs.xml:402]: @WenMeng-NOAA I spoke with @AndrewBenjamin-NOAA about this.  This field is really a column integrated field, but there is no way that I know of to indicate "entire soil column" or something like that.  

I am checking if I can just use one level to indicate simply "CISOILM:0 m underground" similar to the RAP SOILM field.

Comment (EricJames-NOAA) [reply parm/fv3lam_rrfs.xml:402]: @WenMeng-NOAA I confirmed I can output it as CISOILM:0 m underground

Is this approach OK with you?

Comment (WenMeng-NOAA) [reply parm/fv3lam_rrfs.xml:402]: @EricJames-NOAA Your approach works to me since GFS and RRFS use different LSMs.

**Comment by WenMeng-NOAA** [comment]
@EricJames-NOAA My UPP standalone test with RRFS model files from you is located at /home/Wen.Meng/stmp2/rrfs_2024022718 on Hera.
I saw total soil moisture, surface albedo and snow melt generated:
```
1414:8988243285:d=2024022718:CISOILM:0 m underground:6 hour fcst:
1628:10545816268:d=2024022718:ALBDO:surface:6 hour fcst
1446:9165494022:d=2024022718:SNOM:surface:5-6 hour acc fcst
```
Please verify and let me know if you find any inconsistency compared to your test.

**Comment by EricJames-NOAA** [comment]
> @EricJames-NOAA My UPP standalone test with RRFS model files from you is located at /home/Wen.Meng/stmp2/rrfs_2024022718 on Hera. I saw total soil moisture, surface albedo and snow melt generated:
> 
> ```
> 1414:8988243285:d=2024022718:CISOILM:0 m underground:6 hour fcst:
> 1628:10545816268:d=2024022718:ALBDO:surface:6 hour fcst
> 1446:9165494022:d=2024022718:SNOM:surface:5-6 hour acc fcst
> ```
> 
> Please verify and let me know if you find any inconsistency compared to your test.

@WenMeng-NOAA your results look consistent with my tests!  Thanks!

**Comment by WenMeng-NOAA** [comment]
@FernandoAndrade-NOAA You may start the UPP RTs on Hera and Orion. There will be baseline updates for fv3r and 3drtma with this PR. You can refer to my tests on Hera at /home/Wen.Meng/stmp2:
```
Wen.Meng@hfe06 stmp2$ ls -ltr */*diff
-rw-r--r-- 1 Wen.Meng stmp    75 Apr  1 19:29 rap_2020072316_pe_test/WRFPRS.GrbF16.diff
-rw-r--r-- 1 Wen.Meng stmp  2022 Apr  1 19:31 fv3r_2023062800/PRSLEV10.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp  2022 Apr  1 19:31 fv3r_2023062800_pe_test/PRSLEV10.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp 14823 Apr  1 19:32 rtma_2023040400_pe_test/NATLEV00.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp 14823 Apr  1 19:32 rtma_2023040400/NATLEV00.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp 14667 Apr  1 19:33 fv3r_2023062800/NATLEV10.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp  2589 Apr  1 19:34 rtma_2023040400_pe_test/PRSLEV00.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp 14667 Apr  1 19:34 fv3r_2023062800_pe_test/NATLEV10.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp  2589 Apr  1 19:34 rtma_2023040400/PRSLEV00.tm00.diff
```

**Comment by FernandoAndrade-NOAA** [comment]
> @FernandoAndrade-NOAA You may start the UPP RTs on Hera and Orion. There will be baseline updates for fv3r and 3drtma with this PR. You can refer to my tests on Hera at /home/Wen.Meng/stmp2:
> 
> ```
> Wen.Meng@hfe06 stmp2$ ls -ltr */*diff
> -rw-r--r-- 1 Wen.Meng stmp    75 Apr  1 19:29 rap_2020072316_pe_test/WRFPRS.GrbF16.diff
> -rw-r--r-- 1 Wen.Meng stmp  2022 Apr  1 19:31 fv3r_2023062800/PRSLEV10.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp  2022 Apr  1 19:31 fv3r_2023062800_pe_test/PRSLEV10.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp 14823 Apr  1 19:32 rtma_2023040400_pe_test/NATLEV00.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp 14823 Apr  1 19:32 rtma_2023040400/NATLEV00.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp 14667 Apr  1 19:33 fv3r_2023062800/NATLEV10.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp  2589 Apr  1 19:34 rtma_2023040400_pe_test/PRSLEV00.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp 14667 Apr  1 19:34 fv3r_2023062800_pe_test/NATLEV10.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp  2589 Apr  1 19:34 rtma_2023040400/PRSLEV00.tm00.diff
> ```

Hera, Hercules, and Orion RT results match with yours, looks good from my side, thanks.

**Review by FernandoAndrade-NOAA — APPROVED** [review]

**Review by WenMeng-NOAA — APPROVED** [review]

**Comment by WenMeng-NOAA** [comment]
This PR is ready for merging.
